File: sirekap-v.py
import os
import urllib.request
import json
import time
from urllib.parse import urlparse

# Add safe URL opener setup
def create_safe_opener():
    """Create URL opener that only allows HTTP/HTTPS schemes"""
    opener = urllib.request.build_opener(
        urllib.request.HTTPHandler(),
        urllib.request.HTTPSHandler()
    )
    return opener

# Create global safe opener
safe_opener = create_safe_opener()

class ImageDownloader:
    def __init__(self, base_path):
        self.base_path = base_path
        self.opener = safe_opener

    def is_valid_url(self, url):
        """Validate if URL uses HTTPS scheme"""
        parsed = urlparse(url)
        return parsed.scheme == 'https' and parsed.netloc.endswith('kpu.go.id')

    def download_image_with_retry(self, image_url, image_path):
        if not self.is_valid_url(image_url):
            print(f"Invalid URL scheme or domain: {image_url}")
            return

        max_retries = 3
        retries = 0

        while retries < max_retries:
            try:
                # Replace urlretrieve with safe download
                with self.opener.open(image_url) as response:
                    with open(image_path, 'wb') as f:
                        f.write(response.read())
                print(f"Gambar berhasil diunduh dari {image_url}")
                return
            except Exception as e:
                print(f"Error saat mengunduh gambar dari {image_url}: {e}")
                retries += 1

        print(f"Gagal mengunduh gambar setelah {max_retries} percobaan")

    def download_images_sequentially(self, image_urls, folder_path, tps_name):
        # Unduh setiap gambar dalam daftar secara berurutan
        for i, image_url in enumerate(image_urls):
            if not image_url:
                print(f"URL gambar kosong untuk TPS {tps_name}")
                continue

            # Generate nama dan path gambar
            image_extension = os.path.splitext(image_url)[1]
            image_name = f"{tps_name}-{i}{image_extension}"
            image_path = os.path.join(folder_path, image_name)

            try:
                # Coba unduh gambar dengan percobaan
                self.download_image_with_retry(image_url, image_path)
            except Exception as e:
                print(f"Error saat mengunduh gambar dari {image_url}: {e}")

class DataFetcher:
    def __init__(self, base_path):
        self.base_path = base_path
        self.downloader = ImageDownloader(base_path)
        self.base_api_url = "https://sirekap-obj-data.kpu.go.id"
        self.opener = safe_opener

    def is_valid_url(self, url):
        """Validate if URL uses HTTPS scheme"""
        parsed = urlparse(url)
        return parsed.scheme == 'https' and parsed.netloc.endswith('kpu.go.id')

    def safe_urlopen(self, url):
        """Safely open URLs after validation"""
        if not self.is_valid_url(url):
            raise ValueError(f"Invalid URL scheme or domain: {url}")
        return self.opener.open(url)

    def fetch_data(self):
        try:
            api_url = f"{self.base_api_url}/wilayah/pemilu/pdpr/0.json"
            with self.safe_urlopen(api_url) as response:
                provinsi_data = json.loads(response.read())

                for provinsi in provinsi_data:
                    provinsi_folder = os.path.join(self.base_path, provinsi["nama"])
                    os.makedirs(provinsi_folder, exist_ok=True)

                    self.fetch_kabupaten_data(provinsi_folder, provinsi)

        except urllib.error.URLError as e:
            # Penanganan kesalahan URL
            print("URLError: Gagal mengambil data:", e.reason)
            print("Mencoba lagi dalam 5 detik...")
            time.sleep(5)
            self.fetch_data()  # Coba lagi

        except Exception as e:
            # Penanganan kesalahan lainnya
            print("Error saat mengambil data:", e)

        except KeyboardInterrupt:
            # Penanganan interrupt keyboard
            print("Proses dihentikan oleh pengguna.")

    def fetch_kabupaten_data(self, provinsi_folder, provinsi):
        api_url = f"{self.base_api_url}/wilayah/pemilu/ppwp/{provinsi['kode']}.json"
        with self.safe_urlopen(api_url) as response_kabupaten:
            kabupaten_data = json.loads(response_kabupaten.read())

            for kabupaten in kabupaten_data:
                kabupaten_folder = os.path.join(provinsi_folder, kabupaten["nama"])
                os.makedirs(kabupaten_folder, exist_ok=True)

                self.fetch_kecamatan_data(kabupaten_folder, provinsi['kode'], kabupaten)

    def fetch_kecamatan_data(self, kabupaten_folder, provinsi_kode, kabupaten):
        api_url = f"{self.base_api_url}/wilayah/pemilu/ppwp/{provinsi_kode}/{kabupaten['kode']}.json"
        with self.safe_urlopen(api_url) as response_kecamatan:
            kecamatan_data = json.loads(response_kecamatan.read())

            for kecamatan in kecamatan_data:
                kecamatan_folder = os.path.join(kabupaten_folder, kecamatan["nama"])
                os.makedirs(kecamatan_folder, exist_ok=True)

                self.fetch_kelurahan_data(kecamatan_folder, provinsi_kode, kabupaten['kode'], kecamatan)

    def fetch_kelurahan_data(self, kecamatan_folder, provinsi_kode, kabupaten_kode, kecamatan):
        api_url = f"{self.base_api_url}/wilayah/pemilu/ppwp/{provinsi_kode}/{kabupaten_kode}/{kecamatan['kode']}.json"
        with self.safe_urlopen(api_url) as response_kelurahan:
            kelurahan_data = json.loads(response_kelurahan.read())

            for kelurahan in kelurahan_data:
                kelurahan_folder = os.path.join(kecamatan_folder, kelurahan["nama"])
                os.makedirs(kelurahan_folder, exist_ok=True)

                self.fetch_tps_data(kelurahan_folder, provinsi_kode, kabupaten_kode, kecamatan['kode'], kelurahan)

    def fetch_tps_data(self, kelurahan_folder, provinsi_kode, kabupaten_kode, kecamatan_kode, kelurahan):
        api_url = f"{self.base_api_url}/wilayah/pemilu/ppwp/{provinsi_kode}/{kabupaten_kode}/{kecamatan_kode}/{kelurahan['kode']}.json"
        with self.safe_urlopen(api_url) as response_tps:
            tps_data = json.loads(response_tps.read())

            for tps in tps_data:
                detail_url = f"{self.base_api_url}/pemilu/hhcw/pdpr/{provinsi_kode}/{kabupaten_kode}/{kecamatan_kode}/{kelurahan['kode']}/{tps['kode']}.json"
                with self.safe_urlopen(detail_url) as response_tps_detail:
                    tps_detail_data = json.loads(response_tps_detail.read())
                    image_urls = tps_detail_data.get("images", [])
                    self.downloader.download_images_sequentially(image_urls, kelurahan_folder, tps['nama'])

# Penggunaan
base_path = "./C-FORM"
data_fetcher = DataFetcher(base_path)
data_fetcher.fetch_data()

File: sirekap-v.js
const axios = require("axios");
const fs = require("fs");
const path = require("path");

class ImageDownloader {
    constructor(basePath) {
        this.basePath = basePath;
    }

    async downloadImage(imageUrl, imagePath) {
        try {
            // Mengunduh gambar dari URL dan menyimpannya ke file
            const response = await axios.get(imageUrl, {
                responseType: "stream"
            });
            const writer = fs.createWriteStream(imagePath);
            response.data.pipe(writer);
            await new Promise((resolve, reject) => {
                writer.on("finish", resolve);
                writer.on("error", reject);
            });
            console.log(`Image downloaded successfully from ${imageUrl}`);
            return true;
        } catch (error) {
            console.error(`Error downloading image from ${imageUrl}:`, error);
            return false;
        }
    }

    async downloadImagesSequentially(imageUrls, folderPath, tps) {
        for (let i = 0; i < imageUrls.length; i++) {
            const imageUrl = imageUrls[i];
            if (!imageUrl) {
                // Menampilkan pesan jika URL gambar null
                console.error(`Null image URL for TPS ${tps.kode}`);
                continue;
            }
            const imageExtension = path.extname(imageUrl);
            const imageName = `${tps.nama}-${i}${imageExtension}`;
            const imagePath = path.join(folderPath, imageName);
            await this.downloadImage(imageUrl, imagePath);
        }
    }
}

class DataFetcher {
    constructor(basePath) {
        this.basePath = basePath;
        this.imageDownloader = new ImageDownloader(basePath);
    }

    async fetchData() {
        try {
            // Mengambil data pemilu nasional
            const nationalUrl =
                "https://sirekap-obj-data.kpu.go.id/wilayah/pemilu/pdpr/0.json";
            const nationalResponse = await axios.get(nationalUrl);
            const provinces = nationalResponse.data;

            // Loop melalui setiap provinsi
            for (const province of provinces) {
                const provinceFolder = path.join(this.basePath, province.nama);
                // Membuat folder untuk setiap provinsi
                fs.mkdirSync(provinceFolder, {
                    recursive: true
                });
                await this.fetchProvinceData(province, provinceFolder);
            }
        } catch (error) {
            console.error("Error fetching data:", error);
        }
    }

    async fetchProvinceData(province, provinceFolder) {
        try {
            // Mengambil data kabupaten/kota dari suatu provinsi
            const districtUrl = `https://sirekap-obj-data.kpu.go.id/wilayah/pemilu/ppwp/${province.kode}.json`;
            const districtResponse = await axios.get(districtUrl);
            const districts = districtResponse.data;

            // Loop melalui setiap kabupaten/kota
            for (const district of districts) {
                const districtFolder = path.join(provinceFolder, district.nama);
                // Membuat folder untuk setiap kabupaten/kota
                fs.mkdirSync(districtFolder, {
                    recursive: true
                });
                await this.fetchDistrictData(province.kode, district, districtFolder);
            }
        } catch (error) {
            console.error("Error fetching district data:", error);
        }
    }

    async fetchDistrictData(provinceCode, district, districtFolder) {
        try {
            // Mengambil data kecamatan dari suatu kabupaten/kota
            const subDistrictUrl = `https://sirekap-obj-data.kpu.go.id/wilayah/pemilu/ppwp/${provinceCode}/${district.kode}.json`;
            const subDistrictResponse = await axios.get(subDistrictUrl);
            const subDistricts = subDistrictResponse.data;

            // Loop melalui setiap kecamatan
            for (const subDistrict of subDistricts) {
                const subDistrictFolder = path.join(districtFolder, subDistrict.nama);
                // Membuat folder untuk setiap kecamatan
                fs.mkdirSync(subDistrictFolder, {
                    recursive: true
                });
                await this.fetchSubDistrictData(
                    provinceCode,
                    district.kode,
                    subDistrict,
                    subDistrictFolder
                );
            }
        } catch (error) {
            console.error("Error fetching sub-district data:", error);
        }
    }

    async fetchSubDistrictData(
        provinceCode,
        districtCode,
        subDistrict,
        subDistrictFolder
    ) {
        try {
            // Mengambil data desa/kelurahan dari suatu kecamatan
            const villageUrl = `https://sirekap-obj-data.kpu.go.id/wilayah/pemilu/ppwp/${provinceCode}/${districtCode}/${subDistrict.kode}.json`;
            const villageResponse = await axios.get(villageUrl);
            const villages = villageResponse.data;

            // Loop melalui setiap desa/kelurahan
            for (const village of villages) {
                const villageFolder = path.join(subDistrictFolder, village.nama);
                // Membuat folder untuk setiap desa/kelurahan
                fs.mkdirSync(villageFolder, {
                    recursive: true
                });
                await this.fetchVillageData(
                    provinceCode,
                    districtCode,
                    subDistrict.kode,
                    village,
                    villageFolder
                );
            }
        } catch (error) {
            console.error("Error fetching village data:", error);
        }
    }

    async fetchVillageData(
        provinceCode,
        districtCode,
        subDistrictCode,
        village,
        villageFolder
    ) {
        try {
            // Mengambil data TPS (Tempat Pemungutan Suara) dari suatu desa/kelurahan
            const tpsUrl = `https://sirekap-obj-data.kpu.go.id/wilayah/pemilu/ppwp/${provinceCode}/${districtCode}/${subDistrictCode}/${village.kode}.json`;
            const tpsResponse = await axios.get(tpsUrl);
            const tpsData = tpsResponse.data;

            // Loop melalui setiap TPS
            for (const tps of tpsData) {
                // Mengambil detail TPS
                const tpsDetailUrl = `https://sirekap-obj-data.kpu.go.id/pemilu/hhcw/pdpr/${provinceCode}/${districtCode}/${subDistrictCode}/${village.kode}/${tps.kode}.json`;
                const tpsDetailResponse = await axios.get(tpsDetailUrl);
                const tpsDetailData = tpsDetailResponse.data;
                const imageUrls = tpsDetailData.images;

                // Mengunduh gambar-gambar TPS secara berurutan
                await this.imageDownloader.downloadImagesSequentially(
                    imageUrls,
                    villageFolder,
                    tps
                );
            }
        } catch (error) {
            console.error("Error fetching TPS data:", error);
        }
    }
}

// Gunakan kelas DataFetcher untuk memulai pengambilan data dan pengunduhan gambar
const dataFetcher = new DataFetcher("./C-FORM");
dataFetcher.fetchData();
